org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleHeaderPanel.java

```diff
@@ -59,23 +60,45 @@ public void updateRule(AbstractRuleDto ruleInformation) {
     var attribute = ruleInformation.getCleanCodeAttribute().get();
     ruleCleanCodeAttributeLabel.setText(
       clean(attribute.getAttributeCategory().getIssueLabel()) + " | " + clean(attribute.getIssueLabel()));
+    // TODO: Change to SLCORE provided tooltip!;
+    //firstSoftwareQualityLabel.setToolTipText(attribute.getToolTip());
     
     var impacts = ruleInformation.getDefaultImpacts();
     var keys = new ArrayList<SoftwareQuality>(impacts.keySet());
-    firstImpactSeverityIcon.setImage(SonarLintImages.getImpactImage(impacts.get(keys.get(0))));
-    firstSoftwareQualityLabel.setText(clean(keys.get(0).getDisplayLabel()));
+    
+    var firstImpactSeverity = impacts.get(keys.get(0));
+    var firstSoftwareQuality = clean(keys.get(0).getDisplayLabel());
+    firstImpactSeverityIcon.setImage(SonarLintImages.getImpactImage(firstImpactSeverity));
+    firstImpactSeverityIcon.setToolTipText(createToolTip(firstImpactSeverity, firstSoftwareQuality));
+    firstSoftwareQualityLabel.setText(firstSoftwareQuality);
+    // TODO: Change to SLCORE provided tooltip!;
+    //firstSoftwareQualityLabel.setToolTipText(keys.get(0).getToolTip());
     
     if (keys.size() > 1) {
-      secondImpactSeverityIcon.setImage(SonarLintImages.getImpactImage(impacts.get(keys.get(1))));
-      secondSoftwareQualityLabel.setText(clean(impacts.get(keys.get(1)).getDisplayLabel()));
+      var secondImpactSeverity = impacts.get(keys.get(1));
+      var secondSoftwareQuality = clean(keys.get(1).getDisplayLabel());
+      secondImpactSeverityIcon.setImage(SonarLintImages.getImpactImage(secondImpactSeverity));
+      secondImpactSeverityIcon.setToolTipText(createToolTip(secondImpactSeverity, secondSoftwareQuality));
+      secondSoftwareQualityLabel.setText(secondSoftwareQuality);
+      // TODO: Change to SLCORE provided tooltip!;
+      //firstSoftwareQualityLabel.setToolTipText(keys.get(1).getToolTip());
       
       if (keys.size() > 2) {
-        thirdImpactSeverityIcon.setImage(SonarLintImages.getImpactImage(impacts.get(keys.get(2))));
-        thirdSoftwareQualityLabel.setText(clean(impacts.get(keys.get(2)).getDisplayLabel()));
+        var thirdImpactSeverity = impacts.get(keys.get(2));
```

Comment:
The three blocks here have a lot in common, do you think you could factorize them somehow? Maybe have a separate `ImpactPanel` class?

Reply:
Done